pydoctor/astbuilder.py

```diff
@@ -512,36 +485,81 @@ def _annotation_from_attrib(self, expr, ctx):
                 return _infer_type(default)
         return None
 
-    def _unstring_annotation(self, node):
+    def _annotations_from_function(self, func: ast.FunctionDef) -> Mapping[str, ast.expr]:
+        """Get annotations from a function definition.
+        @param func: The function definition's AST.
+        @return: Mapping from argument name to annotation.
+            The name C{return} is used for the return type.
+            Unannotated arguments are omitted.
+        """
+        def _get_all_args() -> Iterator[ast.arg]:
+            base_args = func.args
+            # New on Python 3.8 -- handle absence gracefully
+            try:
+                yield from base_args.posonlyargs
+            except AttributeError:
+                pass
+            yield from base_args.args
+            varargs = base_args.vararg
+            if varargs:
+                yield varargs
+            yield from base_args.kwonlyargs
+            kwargs = base_args.kwarg
+            if kwargs:
+                yield kwargs
+        def _get_all_ast_annotations() -> Iterator[Tuple[str, ast.expr]]:
+            for arg in _get_all_args():
+                if arg.annotation:
+                    yield arg.arg, arg.annotation
+            returns = func.returns
+            if returns:
+                yield 'return', returns
+        return {name: self._unstring_annotation(value)
+                for name, value in _get_all_ast_annotations()}
+
+    def _unstring_annotation(self, node: ast.expr) -> ast.expr:
         """Replace all strings in the given expression by parsed versions.
-        Returns the resulting node, or None if parsing failed.
+        @return: The unstringed node. If parsing fails, an error is logged
+            and the original node is returned.
         """
         try:
-            return _AnnotationStringParser().visit(node)
+            expr = _AnnotationStringParser().visit(node)
+            assert isinstance(expr, ast.expr), expr
+            return expr
         except SyntaxError as ex:
             self.builder.currentMod.report(
                     f'syntax error in annotation: {ex}',
                     lineno_offset=node.lineno)
-            return None
+            return node
 
 
 class _AnnotationStringParser(ast.NodeTransformer):
+    """Implementation of L{ModuleVistor._unstring_annotation()}."""
+
+    def _parse_string(self, value: str) -> ast.expr:
+        stmt, = ast.parse(value).body
+        if isinstance(stmt, ast.Expr):
+            expr = self.visit(stmt.value)
+            assert isinstance(expr, ast.expr), expr
```

Comment:
This seems to be the same assertion in the call above. Why are both necessary?

Also, I think this `assert` depends on the Python parser only returning expressions, and not statements, from parsing annotations. In that case, I don't think this should be an assert, I think we should throw an explicit `SyntaxError`, since this is a problem with the *caller*, not our internal logic.

What do we expect to happen with code like

```
def foo(a: "raise ValueError()"):
     pass
```

This is legal Python code (it will run and produce valid results), but will cause an `AssertionError` here. This does not seem like desirable behavior?

Reply:
> This seems to be the same assertion in the call above. Why are both necessary?

It's a similar assertion, but not redundant, since there are a few intermediate steps inside `NodeTransformer` before the value we return here ends up in `_unstring_annotation()`. Also for mypy we need to have either an assert or cast here to be able to return `expr`. When something is not trivially true, I prefer asserts over casts, since we want to know when theory and practice don't align.

> Also, I think this `assert` depends on the Python parser only returning expressions, and not statements, from parsing annotations. In that case, I don't think this should be an assert, I think we should throw an explicit `SyntaxError`, since this is a problem with the _caller_, not our internal logic.

The `if isinstance(stmt, ast.Expr):` check guarantees that if the top-level node is a statement that is not an expression evaluation, we do in fact raise `SyntaxError`.

> What do we expect to happen with code like
> 
> ```
> def foo(a: "raise ValueError()"):
>      pass
> ```
> 
> This is legal Python code (it will run and produce valid results), but will cause an `AssertionError` here. This does not seem like desirable behavior?

This will result in `isinstance(stmt, ast.Expr)` being false and therefore `SyntaxError` will be raised. The `test_bad_string_annotation()` test case checks this, when `pass` is used as the annotation.

This was in fact an error that existed before, but caught because mypy forced me to check everywhere that the AST nodes we expect to be expressions can only be expressions, even on invalid inputs.